cmd/xl-v1-multipart-common.go

```diff
@@ -20,169 +20,147 @@ import (
+// described in uCh.
+func (xl xlObjects) writeUploadJSON(bucket, object string, uCh uploadIDChange) error {
 	uploadsPath := path.Join(mpartMetaPrefix, bucket, object, uploadsJSONFile)
 	uniqueID := getUUID()
 	tmpUploadsPath := path.Join(tmpMetaPrefix, uniqueID)
-	var errs = make([]error, len(xl.storageDisks))
-	var wg = &sync.WaitGroup{}
 
-	// Update `uploads.json` for all the disks.
-	for index, disk := range xl.storageDisks {
+	// slice to store errors from disks
+	errs := make([]error, len(xl.storageDisks))
+	// slice to store if it is a delete operation on a disk
+	isDelete := make([]bool, len(xl.storageDisks))
+
+	wg := sync.WaitGroup{}
+	diskPermutation := xl.getLoadBalancedDisks()
+	for index, disk := range diskPermutation {
 		if disk == nil {
-			errs[index] = errDiskNotFound
+			errs[index] = traceError(errDiskNotFound)
 			continue
 		}
+		// Update `uploads.json` in a go routine.
 		wg.Add(1)
-		// Update `uploads.json` in routine.
 		go func(index int, disk StorageAPI) {
 			defer wg.Done()
-			uploadsBytes, wErr := json.Marshal(uploadsJSON)
-			if wErr != nil {
-				errs[index] = traceError(wErr)
-				return
+
+			// read and parse uploads.json on this disk
+			uploadsJSON, err := readUploadsJSON(bucket, object, disk)
+			if errorCause(err) == errFileNotFound {
+				// If file is not found, we assume an
+				// default (empty) upload info.
+				uploadsJSON, err = newUploadsV1("xl"), nil
 			}
-			if wErr = disk.AppendFile(minioMetaBucket, tmpUploadsPath, uploadsBytes); wErr != nil {
-				errs[index] = traceError(wErr)
+			// If we have a read error, we store error and
+			// exit.
+			if err != nil {
+				errs[index] = traceError(err)
 				return
 			}
-			if wErr = disk.RenameFile(minioMetaBucket, tmpUploadsPath, minioMetaBucket, uploadsPath); wErr != nil {
-				errs[index] = traceError(wErr)
-				return
+
+			if !uCh.isRemove {
+				// Add the uploadID
+				uploadsJSON.AddUploadID(
+					uCh.uploadID,
+					uCh.initiated,
+				)
+			} else {
+				// Remove the upload ID
+				uploadsJSON.RemoveUploadID(uCh.uploadID)
+				if len(uploadsJSON.Uploads) == 0 {
+					isDelete[index] = true
+				}
+			}
+
+			// Check if there are entries to write.
+			if !isDelete[index] {
+				// Serialize to prepare to write to disk.
+				uploadsJSONBytes, wErr := json.Marshal(&uploadsJSON)
+				if wErr != nil {
+					errs[index] = traceError(wErr)
+					return
+				}
+
+				// Write `uploads.json` to disk. First to tmp
+				// location and then rename.
+				if wErr = disk.AppendFile(minioMetaBucket, tmpUploadsPath, uploadsJSONBytes); wErr != nil {
+					errs[index] = traceError(wErr)
+					return
+				}
+				wErr = disk.RenameFile(minioMetaBucket, tmpUploadsPath, minioMetaBucket, uploadsPath)
+				if wErr != nil {
+					if dErr := disk.DeleteFile(minioMetaBucket, tmpUploadsPath); dErr != nil {
+						// we store the most recent error.
+						errs[index] = traceError(dErr)
+						return
+					}
+					errs[index] = traceError(wErr)
```

Comment:
This whole code here can be moved to a single function. Its common across FS and XL.